File: packages/node/src/cachelint/report.ts
/**
 * Plain-text report renderer (78-column, deterministic) — byte-identical to
 * the Python renderer for any analyze() result (the CLI cross-diff test
 * compares the two directly). Wrapping measures code points (SPEC.md §1.1);
 * thousands separators are locale-independent (SPEC.md §1.5).
 */

import { cpLength } from './util/codepoints.js';
import { formatFixed } from './util/halfEven.js';
import { formatThousands } from './util/pyformat.js';

const W = 78;
const SEV_TAG: Record<string, string> = { high: '[HIGH]', medium: '[MED ]', low: '[low ]' };

const VERDICT_LABEL: Record<string, string> = {
  CACHEABLE: 'CACHEABLE',
  CACHEABLE_WITH_ACTION: 'CACHEABLE (action required)',
  CACHEABLE_PREFIX: 'CACHEABLE PREFIX (dynamic tail)',
  BELOW_THRESHOLD: "BELOW THRESHOLD (won't cache)",
  EFFECTIVELY_UNCACHEABLE: 'EFFECTIVELY UNCACHEABLE',
  UNKNOWN_THRESHOLD: 'UNKNOWN THRESHOLD (structural analysis only)',
};

function rule(char = '-'): string {
  return char.repeat(W);
}

function wrap(text: string, indent = 4, width = W): string[] {
  const pad = ' '.repeat(indent);
  const out: string[] = [];
  let cur = '';
  for (const w of text.split(/\s+/).filter(Boolean)) {
    const candidate = cur ? cur + ' ' + w : w;
    if (pad.length + cpLength(candidate) > width && cur) {
      out.push(pad + cur);
      cur = w;
    } else {
      cur = candidate;
    }
  }
  if (cur) out.push(pad + cur);
  return out;
}

function fmtTokens(n: unknown): string {
  return typeof n === 'number' && Number.isInteger(n) ? `~${formatThousands(n)}` : '?';
}

type Dict = Record<string, any>;

export function render(result: Dict): string {
  const lines: string[] = [];
  lines.push(rule('='));
  lines.push(' CACHELINT - prompt-cache pre-flight report');
  lines.push(`   knowledge base: ${result.meta.knowledge_base}`);
  lines.push(rule('='));

  for (const item of result.items as Dict[]) {
    lines.push('');
    let header = ` INPUT #${item.index} -- ${item.provider} / ${item.model || '(no model)'}`;
    if (item.upstream) header += `  [upstream: ${item.upstream}]`;
    lines.push(header);
    lines.push(rule());

    const v = item.verdict;
    lines.push(` Verdict   : ${VERDICT_LABEL[v.code] ?? v.code}`);
    lines.push(...wrap(v.summary, 13));

    const tr = item.token_report;
    const total = tr.total;
    const secs = Object.entries(tr.sections as Dict)
      .map(([name, n]) => `${name}=${fmtTokens(n)}`)
      .join('  ');
    lines.push(` Tokens    : total ${fmtTokens(total.tokens)}  (${total.method})`);
    if (secs) lines.push(...wrap(secs, 13));

    const th = item.threshold;
    if (th.min_tokens !== null) {
      const meets = th.meets ? 'meets' : 'BELOW';
      const inc = th.increment ? `, +${th.increment}-tok steps` : '';
      lines.push(
        ` Threshold : ${formatThousands(th.min_tokens)} min (${th.tier}${inc}) -> prompt ${meets} the floor`,
      );
    } else {
      lines.push(` Threshold : unpublished (${th.tier})`);
    }

    const sr = item.system_report;
    if (sr.present) {
      let unitOk = '';
      if (th.min_tokens !== null) {
        unitOk = sr.cached_unit_meets_threshold
          ? ' -> clears the floor alone'
          : ' -> below the floor alone';
      }
      lines.push(
        ` SystemRpt : system ${fmtTokens(sr.tokens)} tok | cached unit ` +
          `(tools+system) ${fmtTokens(sr.cached_unit_tokens)} tok${unitOk}` +
          (sr.findings_in_system ? ` | ${sr.findings_in_system} finding(s) inside system` : ''),
      );
    }

    const findings = item.findings as Dict[];
    if (findings.length) {
      lines.push(` Findings  : ${findings.length}`);
      for (const f of findings.slice(0, 10)) {
        const tag = SEV_TAG[f.severity] ?? `[${f.severity}]`;
        lines.push(
          `   ${tag} ${f.kind}  @ ${f.section}+${f.section_offset}` +
            `  (${formatFixed(f.pct_into_prompt, 0)}% into prompt)`,
        );
        lines.push(...wrap(`"${f.excerpt}"`, 10));
        lines.push(...wrap('-> ' + f.why, 10));
      }
      if (findings.length > 10) {
        lines.push(`   ... and ${findings.length - 10} more`);
      }
    } else {
      lines.push(' Findings  : none — no dynamic-content indicators detected');
    }

    const sp = item.stable_prefix;
    const lim = sp.limited_by;
    if (lim) {
      lines.push(
        ` StablePfx : ${fmtTokens(sp.tokens)} tok ` +
          `(${sp.pct_of_prompt}% of prompt), capped by ${lim.kind} ` +
          `at ${lim.section}+${lim.section_offset}`,
      );
      if (sp.if_first_fixed_tokens !== null && sp.if_first_fixed_tokens !== undefined) {
        lines.push(
          `             fixing that first finding alone -> ` +
            `${fmtTokens(sp.if_first_fixed_tokens)} tok`,
        );
      }
    } else {
      lines.push(` StablePfx : ${fmtTokens(sp.tokens)} tok (entire prompt is stable)`);
    }

    const en = item.enablement;
    lines.push(` Enablement: ${(en.mode as string).toUpperCase()}`);
    lines.push(...wrap(en.detail, 13));

    if ((item.recommendations as string[]).length) {
      lines.push(' Recommendations:');
      (item.recommendations as string[]).forEach((rec, i) => {
        lines.push(...wrap(`${i + 1}. ${rec}`, 4));
      });
    }

    const intel = item.provider_intel;
    lines.push(' Intel     :');
    lines.push(...wrap(`TTL: ${intel.ttl}`, 6));
    lines.push(...wrap(`Read discount: ${intel.read_discount}`, 6));
    lines.push(...wrap(`Write cost: ${intel.write_cost}`, 6));
    lines.push(...wrap(`Verify via: ${intel.verify_field}`, 6));
    for (const g of intel.gotchas as string[]) {
      lines.push(...wrap(`! ${g}`, 6));
    }

    for (const w of (item.warnings as string[]) ?? []) {
      lines.push(...wrap(`(warning) ${w}`, 2));
    }
  }

  for (const seq of (result.sequences as Dict[]) ?? []) {
    const g = seq.group;
    lines.push('');
    if (seq.count === 0) {
      // cross-group note
      lines.push(rule('='));
      lines.push(...wrap(seq.summary, 1));
      continue;
    }
    lines.push(
      ` SEQUENCE -- ${g.provider} / ${g.model} ` +
        `(${seq.count} request(s): #${(seq.indices as number[]).join(', #')})`,
    );
    lines.push(rule());
    if (seq.shared_prefix_all) {
      const spa = seq.shared_prefix_all;
      const ok = spa.meets_threshold ? 'clears' : 'does NOT clear';
      const floor = spa.min_tokens !== null ? formatThousands(spa.min_tokens) : 'unpublished';
      lines.push(
        ` Shared prefix across all: ${fmtTokens(spa.tokens)} tok ` + `-> ${ok} the floor (${floor})`,
      );
    }
    for (const p of seq.pairs as Dict[]) {
      lines.push(
        ` #${p.from} -> #${p.to}  ${(p.expected as string).padEnd(8)} ` +
          `relation=${p.relation}  common prefix ${fmtTokens(p.lcp_tokens)} tok` +
          (p.expected_cached_tokens ? ` (cached ${fmtTokens(p.expected_cached_tokens)})` : ''),
      );
      const d = p.diverged_at;
      if (d) {
        lines.push(`      diverged at ${d.section}+${d.section_offset} ` + `(char ${d.offset})`);
        if ((p.likely_causes as string[]).length) {
          lines.push(...wrap('likely cause: ' + (p.likely_causes as string[]).join('; '), 6));
        }
        lines.push(...wrap(`prev: "${d.context_prev}"`, 6));
        lines.push(...wrap(`curr: "${d.context_curr}"`, 6));
      }
      const t = p.ttl_check;
      if (t) {
        let gapLine = `gap between requests: ${t.gap_seconds}s`;
        if (t.note) gapLine += ` — ${t.note}`;
        lines.push(...wrap(gapLine, 6));
      }
      for (const n of p.notes as string[]) {
        lines.push(...wrap('note: ' + n, 6));
      }
    }
    lines.push(...wrap('Summary: ' + seq.summary, 1));
  }

  lines.push('');
  lines.push(rule('='));
  lines.push(...wrap('Caveat: ' + result.meta.caveat, 1));
  return lines.join('\n');
}

File: packages/python/tryaii/cachelint/report.py
"""Plain-text report renderer (78-column, deterministic).

Byte-identical to the Node SDK's report.ts for any normalized analyze()
result — the CLI cross-diff test compares the two directly. Numbers arrive
already normalized (SPEC.md §1.3), so str()/template interpolation agrees
across languages; thousands separators use the shared manual format.
"""

from __future__ import annotations

_W = 78
_SEV_TAG = {"high": "[HIGH]", "medium": "[MED ]", "low": "[low ]"}

_VERDICT_LABEL = {
    "CACHEABLE": "CACHEABLE",
    "CACHEABLE_WITH_ACTION": "CACHEABLE (action required)",
    "CACHEABLE_PREFIX": "CACHEABLE PREFIX (dynamic tail)",
    "BELOW_THRESHOLD": "BELOW THRESHOLD (won't cache)",
    "EFFECTIVELY_UNCACHEABLE": "EFFECTIVELY UNCACHEABLE",
    "UNKNOWN_THRESHOLD": "UNKNOWN THRESHOLD (structural analysis only)",
}


def _rule(char: str = "-") -> str:
    return char * _W


def _wrap(text: str, indent: int = 4, width: int = _W) -> list[str]:
    pad = " " * indent
    out, cur = [], ""
    for w in text.split():
        candidate = (cur + " " + w) if cur else w
        if len(pad) + len(candidate) > width and cur:
            out.append(pad + cur)
            cur = w
        else:
            cur = candidate
    if cur:
        out.append(pad + cur)
    return out


def _fmt_tokens(n) -> str:
    return f"~{n:,}" if isinstance(n, int) and not isinstance(n, bool) else "?"


def render(result: dict) -> str:
    lines: list[str] = []
    lines.append(_rule("="))
    lines.append(" CACHELINT - prompt-cache pre-flight report")
    lines.append(f"   knowledge base: {result['meta']['knowledge_base']}")
    lines.append(_rule("="))

    for item in result["items"]:
        lines.append("")
        header = f" INPUT #{item['index']} -- {item['provider']} / {item['model'] or '(no model)'}"
        if item.get("upstream"):
            header += f"  [upstream: {item['upstream']}]"
        lines.append(header)
        lines.append(_rule())

        v = item["verdict"]
        lines.append(f" Verdict   : {_VERDICT_LABEL.get(v['code'], v['code'])}")
        lines.extend(_wrap(v["summary"], indent=13))

        tr = item["token_report"]
        total = tr["total"]
        secs = "  ".join(f"{name}={_fmt_tokens(n)}" for name, n in tr["sections"].items())
        lines.append(f" Tokens    : total {_fmt_tokens(total['tokens'])}  ({total['method']})")
        if secs:
            lines.extend(_wrap(secs, indent=13))

        th = item["threshold"]
        if th["min_tokens"] is not None:
            meets = "meets" if th["meets"] else "BELOW"
            inc = f", +{th['increment']}-tok steps" if th.get("increment") else ""
            lines.append(
                f" Threshold : {th['min_tokens']:,} min ({th['tier']}{inc}) -> prompt {meets} the floor")
        else:
            lines.append(f" Threshold : unpublished ({th['tier']})")

        sr = item["system_report"]
        if sr["present"]:
            unit_ok = ""
            if th["min_tokens"] is not None:
                unit_ok = (" -> clears the floor alone" if sr["cached_unit_meets_threshold"]
                           else " -> below the floor alone")
            lines.append(f" SystemRpt : system {_fmt_tokens(sr['tokens'])} tok | cached unit "
                         f"(tools+system) {_fmt_tokens(sr['cached_unit_tokens'])} tok{unit_ok}"
                         + (f" | {sr['findings_in_system']} finding(s) inside system"
                            if sr["findings_in_system"] else ""))

        findings = item["findings"]
        if findings:
            lines.append(f" Findings  : {len(findings)}")
            for f in findings[:10]:
                tag = _SEV_TAG.get(f["severity"], f"[{f['severity']}]")
                lines.append(f"   {tag} {f['kind']}  @ {f['section']}+{f['section_offset']}"
                             f"  ({f['pct_into_prompt']:.0f}% into prompt)")
                lines.extend(_wrap(f'"{f["excerpt"]}"', indent=10))
                lines.extend(_wrap("-> " + f["why"], indent=10))
            if len(findings) > 10:
                lines.append(f"   ... and {len(findings) - 10} more")
        else:
            lines.append(" Findings  : none — no dynamic-content indicators detected")

        sp = item["stable_prefix"]
        lim = sp.get("limited_by")
        if lim:
            lines.append(f" StablePfx : {_fmt_tokens(sp['tokens'])} tok "
                         f"({sp['pct_of_prompt']}% of prompt), capped by {lim['kind']} "
                         f"at {lim['section']}+{lim['section_offset']}")
            if sp.get("if_first_fixed_tokens") is not None:
                lines.append(f"             fixing that first finding alone -> "
                             f"{_fmt_tokens(sp['if_first_fixed_tokens'])} tok")
        else:
            lines.append(f" StablePfx : {_fmt_tokens(sp['tokens'])} tok (entire prompt is stable)")

        en = item["enablement"]
        lines.append(f" Enablement: {en['mode'].upper()}")
        lines.extend(_wrap(en["detail"], indent=13))

        if item["recommendations"]:
            lines.append(" Recommendations:")
            for i, rec in enumerate(item["recommendations"], 1):
                lines.extend(_wrap(f"{i}. {rec}", indent=4))

        intel = item["provider_intel"]
        lines.append(" Intel     :")
        lines.extend(_wrap(f"TTL: {intel['ttl']}", indent=6))
        lines.extend(_wrap(f"Read discount: {intel['read_discount']}", indent=6))
        lines.extend(_wrap(f"Write cost: {intel['write_cost']}", indent=6))
        lines.extend(_wrap(f"Verify via: {intel['verify_field']}", indent=6))
        for g in intel["gotchas"]:
            lines.extend(_wrap(f"! {g}", indent=6))

        for w in item.get("warnings", []):
            lines.extend(_wrap(f"(warning) {w}", indent=2))

    for seq in result.get("sequences", []):
        g = seq["group"]
        lines.append("")
        if seq["count"] == 0:   # cross-group note
            lines.append(_rule("="))
            lines.extend(_wrap(seq["summary"], indent=1))
            continue
        lines.append(f" SEQUENCE -- {g['provider']} / {g['model']} "
                     f"({seq['count']} request(s): #{', #'.join(map(str, seq['indices']))})")
        lines.append(_rule())
        if seq.get("shared_prefix_all"):
            spa = seq["shared_prefix_all"]
            ok = "clears" if spa["meets_threshold"] else "does NOT clear"
            floor = f"{spa['min_tokens']:,}" if spa["min_tokens"] is not None else "unpublished"
            lines.append(f" Shared prefix across all: {_fmt_tokens(spa['tokens'])} tok "
                         f"-> {ok} the floor ({floor})")
        for p in seq["pairs"]:
            lines.append(f" #{p['from']} -> #{p['to']}  {p['expected']:<8} "
                         f"relation={p['relation']}  common prefix {_fmt_tokens(p['lcp_tokens'])} tok"
                         + (f" (cached {_fmt_tokens(p['expected_cached_tokens'])})"
                            if p["expected_cached_tokens"] else ""))
            d = p.get("diverged_at")
            if d:
                lines.append(f"      diverged at {d['section']}+{d['section_offset']} "
                             f"(char {d['offset']})")
                if p["likely_causes"]:
                    lines.extend(_wrap("likely cause: " + "; ".join(p["likely_causes"]), indent=6))
                lines.extend(_wrap(f'prev: "{d["context_prev"]}"', indent=6))
                lines.extend(_wrap(f'curr: "{d["context_curr"]}"', indent=6))
            t = p.get("ttl_check")
            if t:
                gap_line = f"gap between requests: {t['gap_seconds']}s"
                if t.get("note"):
                    gap_line += f" — {t['note']}"
                lines.extend(_wrap(gap_line, indent=6))
            for n in p["notes"]:
                lines.extend(_wrap("note: " + n, indent=6))
        lines.extend(_wrap("Summary: " + seq["summary"], indent=1))

    lines.append("")
    lines.append(_rule("="))
    lines.extend(_wrap("Caveat: " + result["meta"]["caveat"], indent=1))
    return "\n".join(lines)
